src/DbPostgreSQLAdapter/Database.cpp

```diff
@@ -0,0 +1,134 @@
+#include "stdafx.h"
+
+#include "Database.h"
+#include "DbAdapterInterface/ITable.h"
+#include "RecordSet.h"
+#include "Table.h"
+#include "TableRecordSet.h"
+#include "Transaction.h"
+
+#include <libpq-fe.h>
+
+namespace {
+	std::unique_ptr<PGresult, void(*)(PGresult*)> createRAIIPGresult(PGresult* result)
+	{
+		return std::unique_ptr<PGresult, void(*)(PGresult*)>(result, PQclear);
+	}
+}
+
+namespace systelab::db::postgresql {
+
+	Database::Database(PGconn* database)
+		: m_database(database)
+		, m_lastOperationRowsAffected(0)
+		, m_lastInsertedRowId(0)
+	{
+	}
+
+	Database::~Database()
+	{
+		PQfinish(m_database);
+	}
+
+	ITable& Database::getTable(std::string tableName)
+	{
+		const auto itr = m_tables.find(tableName);
+		if (itr == m_tables.cend())
+		{
+			m_tables.insert(std::make_pair(tableName, new Table(*this, tableName)));
+		}
+
+		const auto& table = m_tables.at(tableName);
+		return *(table.get());
+	}
+
+	std::unique_ptr<IRecordSet> Database::executeQuery(const std::string& query)
+	{
+		std::lock_guard<std::recursive_mutex> lock(m_mutex);
+		const auto statementResult = createRAIIPGresult(PQexec(m_database, query.data()));
+		if (PQresultStatus(statementResult.get()) == PGRES_TUPLES_OK)
+		{
+			return std::make_unique<RecordSet>(statementResult.get());
+		}
+		else
+		{
+			const std::string errorMessage = PQresultErrorMessage(statementResult.get());
+			
+			std::ostringstream exceptionStrem;
+			exceptionStrem << "# ERR: SQLException in " << __FILE__;
+			exceptionStrem << "(Database::executeQuery) on line " << __LINE__ << '\n';
+			exceptionStrem << "# ERR: " << errorMessage << std::endl;
+
+			throw std::runtime_error(exceptionStrem.str());
+		}
+	}
+
+	std::unique_ptr<ITableRecordSet> Database::executeTableQuery(const std::string& query, ITable& table)
+	{	
+		std::lock_guard<std::recursive_mutex> lock(m_mutex);
+		const auto statementResult = createRAIIPGresult(PQexec(m_database, query.data()));
+		if (PQresultStatus(statementResult.get()) == PGRES_TUPLES_OK)
+		{
+			return std::make_unique<TableRecordSet>(table, statementResult.get());
+		}
+		else
+		{
+			const std::string errorMessage = PQresultErrorMessage(statementResult.get());
```

Comment:
This is claiming for a generic function to build and throw error, probably like:
```
void throwPostgressException(const PGResult*, const std::source_location& srcLocation = std::source_location::current())
{
    const std::string errorMessage = PQresultErrorMessage(statementResult.get());
    std::ostringstream exceptionStrem;
    exceptionStrem << "# ERR: SQLException in " << srcLocation.file_name()
                              << "(Database::executeOperation) on line " << srcLocation.line() << std::endl
                              << "# ERR: " << errorMessage << std::endl;
throw std::runtime_error(exceptionStrem.str()); // Or specific exception if available
}
```
Using it as:
```
throwPostgressException(statementResult.get());
```

Reply:
Granted, you only have two more wishes left xD